### PR TITLE
Asyncify Recursively

### PR DESCRIPTION
**Pull request by FreeMasen**
As illustrated in the new asyncify test, we are not currently recursively swapping socket for cosock.socket. Only the top level module is getting swapped.

### CODE REVIEW

File: cosock.lua [hunk 1]
```diff
@@ -133,18 +135,17 @@ do
...
     elseif name == "ssl.https" then
-      name = "cosock.ssl.https"
+      subbed = "cosock.ssl.https"
     elseif name == "ssl" then
-      name = "cosock.ssl"
+      subbed = "cosock.ssl"
     end
-
-    return realrequire(name)
+    return subbed and realrequire(subbed) or m.asyncify(name)
```
Comment (FreeMasen) [inline cosock.lua:148]: My initial thought was that just calling `realrequire("cosock.socket")` here should have been enough, however this regresses the issues addressed in #22, I am not sure how we are calling `wrappedrequire` but not able to attach that to the usage of `ssl.https`.

Comment (FreeMasen) [reply cosock.lua:148]: The issue comes down to the behavior of requiring the ssl module. We expect the pass through definitions from cosock's ssl module but would get the non pass through version. This if statement can be reduced to just consider socket and ssl which should fix our issue without needing the manually wrapped modules like cosock.ssl.https

File: cosock.lua [hunk 2]
```diff
@@ -174,8 +173,13 @@ do
...
         -- load module with loader
         local module = loader(name, loc)
-
+        package.asyncify_loaded[name] = module
         return module
+      -- If the last searcher returns nil we need to check the
+      -- package.loaded or we may miss some std lib packages
+      elseif loader == nil and package.loaded[name] then
+        package.asyncify_loaded[name] = package.loaded[name]
+        return package.loaded[name]
```
Comment (azdle) [inline cosock.lua:182]: I'm not sure this is 100% right still.

There's a few corner cases like

```
require 'foo' --< requires 'socket'
cosock.asyncify 'foo`
```

I think that the asyncify call will just end up getting the non-asyncified version of 'foo' out of package.loaded (which should be the first searcher and can't itself be asyncified).

If you wanted to add some more super strict tests that verify edge cases like that it would help be be satisfied this all works quicker, otherwise I'll get to it myself at some point.

Comment (FreeMasen) [reply cosock.lua:182]: I've updated the tests to determine that requiring the two tests targets do not return `cosock.socket` and `cosock.asyncify` isn't impacted by those requires.

Comment (FreeMasen) [reply cosock.lua:182]: https://github.com/FreeMasen/cosock/blob/fix/recursive-asyncify/test/asyncify/asyncify-works.lua#L53-L54

**Comment by FreeMasen** [comment]
The last commit includes a change to allow for calling `cosock.asyncify "socket"` and get back `cosock.socket`. While this isn't something that I expect to happen all that often, since we provide the properties on the `cosock` object already, I figured it was counter intuitive that calling `cosock.asyncify "socket"` would return the Luasocket module.

**Comment by FreeMasen** [comment]
An alternative to the `package.loaded` last would be to potentially pre-populated `package.asyncify_loaded` with the modules defined [here](https://github.com/lua/lua/blob/eadd8c7178c79c814ecca9652973a9b9dd4cc71b/linit.c#L42-L54)

**Comment by FreeMasen** [comment]
Here is an alternative approach to the remaining question for this PR

https://github.com/FreeMasen/cosock/pull/1

File: cosock.lua [hunk 3]
```diff
@@ -59,6 +59,8 @@ local print = function() end
 
 local m = {}
 
+m._VERSION = "0.2.0"
+socket._VERSION = m._VERSION
```
Comment (azdle) [inline cosock.lua:63]: IIRC, this is overriding https://github.com/cosock/cosock/pull/24/files#diff-cfc562be408a5546f486f3ba8fb25b63fb31b83313a49278690fdc084760434fR14

Should this be `socket._COSOCK_VERSION = m._VERSION`? (Which would allow https://github.com/cosock/cosock/pull/24/files#diff-cfc562be408a5546f486f3ba8fb25b63fb31b83313a49278690fdc084760434fR17 to be removed so we don't have to update it in two places.)

Comment (FreeMasen) [reply cosock.lua:63]: Good call, I was hoping that would be an option but for some reason kept thinking I'd need to `require('cosock')` in `socket.lua` to make that work. This is much better

File: cosock.lua [hunk 4]
```diff
@@ -154,6 +151,12 @@ do
 
   function m.asyncify(name)
     local err
+    if name == "socket" then name = "cosock.socket" end
+    if name == "ssl" then name = "cosock.ssl" end
+    package.asyncify_loaded = package.asyncify_loaded or {}
+    if package.asyncify_loaded[name] then
+      return package.asyncify_loaded[name]
+    end
```
Comment (azdle) [inline cosock.lua:159]: The last thing I'm still waffling on is the naming & placement of `asyncify_loaded`. Would you be okay with changing this to a local variable within the asyncify do block? Then it's not part of any public API and I can think about where it should be exposed later.

**Review by azdle — APPROVED** [review]
Approved pending thew two comments above and getting the `fixup` commits squashed.

Sorry it has taken me so long to get this reviewed. And thanks for this fix!

File: cosock.lua (hunk 3 above)
Comment (FreeMasen) [reply cosock.lua:63]: Actually, this would end up missing that property when calling `require "cosock.socket"` since it would only end up being populated on `require "cosock"`. 

This is mostly here to be able to test that we have the cosock.socket module over luasocket module. One option would be to centralize the version in a `version.lua` file, which would just return the version string and could be imported into any of the sub-modules. 

Another way to go would be to remove this property entirely and adjust the tests to use table comparison over checking this property. Essentially, now that we are doing [this](https://github.com/cosock/cosock/pull/24/files#diff-4fb91077f8627b1858d0faf4936f70e0d875f4a29f6144712251bfab24cdf499R53) I think the `assert` above that line is no longer needed

Comment (FreeMasen) [reply cosock.lua]: Ok, this ended up being a bit of a mess to sort out. [I just pushed an update that](https://github.com/cosock/cosock/pull/24/files#diff-4fb91077f8627b1858d0faf4936f70e0d875f4a29f6144712251bfab24cdf499R9-R26) streamlines each module we are testing. It validates that both the sync and async versions are the ones we expect.

It does this by calling `socket.select` from each of the modules, expecting that luasocket will timeout and that cosock will raise an error (yielding from outside of a co-routine). This allows us to not have to populate a `_COSOCK_VERSION` on any sub-modules but still validate we are getting the module we were expecting. Let me know if you have any objections to that test's validity

Comment (azdle) [reply cosock.lua]: That's clever, I like it. No objections to using that.

**Comment by azdle** [comment]
I made a couple of small (pedantic) changes, squashing your almost empty 3rd commit (https://github.com/cosock/cosock/pull/24/commits/fefbe8af621f1ab6d0d1c307f200458e3db0868e) into the first commit and removing the `socket._VERSON` that was added in the first commit and then removed in the second.

Then I accidentally force pushed to you branch. :/ Sorry. Hope you're okay with those changes.

This should be all good to merge now.

File: cosock.lua [hunk 5]
```diff
@@ -59,6 +59,8 @@ local print = function() end
 
 local m = {}
 
+m._VERSION = "0.2.0"
+socket._COSOCK_VERSION = m._VERSION
```
[the remark this answers is not in the document]
Comment (azdle) [reply cosock.lua]: Hmm, I think it's something that should exist. But yeah, the inconsistency definitely makes it less useful. I'm cool with dropping it here. I'll open an issue to remind myself to decide on a way to add that that's consistent.
